### PR TITLE
[PR] Reorganizing and enhancing document 

### LINKED ISSUE #46: Enhance: Update + Section on Perf Testing
@LuchoTurtle please read through this learning doc and comprehensively update it. 🙏 
My understanding is that you have _limited_ experience with **_`public`_ APIs**. 💭 
Please _list_ the APIs you _have_ used in a comment ***`before`*** you start work on this. 📝 
Our objective is to capture as much knowledge as we can so that we can build the _best_ possible API for our 
[**`SaaS` Product**](https://github.com/dwyl/app/issues/273) so we need to do first-principals research/analysis of the best _existing_ examples we can learn from. 👌 

# Todo:

+ [x] Read the `README.md` with a [beginner's mind](https://en.wikipedia.org/wiki/Shoshin) i.e. as if you were ***`new`*** to all this. 👀 
+ [x] Go through all the links in the `README.md` and determine their relevance. 🔗 
  + [x] If any links are no longer relevant, please drop a comment in this thread and _remove_ the link. ✂️ 
+ [x] Find links to the best **Public `REST` + `Realtime` APIs**. Some Googling/Research required. 🔍 
  + [x] Show _example_ request/responses where applicable. 🧑‍💻 

+ [x] Add ***`new`*** section on **_performance_ and _reliability_ testing** for APIs.
  e.g: imagine we are building an API for our **`SaaS` Product** and want to run a series of performance tests each time a commit is made and pushed on a PR. i.e. on GitHub Actions. What is required for that?
  + [x] Second, how do we _monitor_ the "health" of an API so we know _immediately_ when it has degraded performance?

This is at _least_ a day worth of reading, research and write-up. ⏳ 
We need to make this investment up-front so that we can make the best possible API we can. 🙏

### PR DESCRIPTION
**Pull request by LuchoTurtle**
closes #46

This PR will effectively reorganize the document, make it more structure and add sections that delve into more detail according to API design and best practices.

### CODE REVIEW

**Comment by LuchoTurtle** [comment]
Took a little longer to delve deeper into versioning and adding best API practices/standardization. 

I'm close to finishing this and will be done by Monday for sure. I want to give a comprehensive guide to documenting and doing stress/performance tests on APIs (still need to do more research for the best options). 

Pruned links, added new ones and deleted resources that were outdated or outright 404'd.

**Comment by LuchoTurtle** [comment]
The refactoring of the whole document should now be complete.

I tried to maintain the pre-existing content (it was impossible sometimes, a few links were 404) and added new content according to #46 .

I initially intended to add more but I think having sections concise is a much easier read.

Let me know if there is something missing/ if I should elaborate more in any topic.

**Comment by nelsonic** [comment]
Thanks @LuchoTurtle will go through this PR next. 👍

File: README.md [hunk 1]
```diff
@@ -1231,6 +1231,11 @@ Besides managing different collaboratively APIs,
 creating Open API specifications,
 documentation
 and monitoring.
+- [**Insomnia**](https://insomnia.rest/):
+similar to Postman, it's a API platform
```
Comment (nelsonic) [inline README.md:1235]:
```suggestion
similar to Postman, but ***`Open Source`***,  is an `App`
```

**Comment by nelsonic** [comment]
Picking up the review on this _now_ as I've seen a couple of links to this in https://github.com/dwyl/mvp/pull/263/files#r1070223004 and rather than having to update the link:

```
https://github.com/dwyl/learn-api-design/blob/revamp/README.md#show-meaningful-errors
```
Would be better if it linked directly to the `main` branch.

So this is my side quest for the next hour. 🧑‍💻 👀 ⏳

File: README.md [hunk 2]
```diff
@@ -56,50 +73,187 @@ and [Kevin Lacker](https://twitter.com/lacker)'s ([@Parse](http://parseplatform.
...
+<img width="1418" alt="workflow RPC" src="https://user-images.githubusercontent.com/17494745/209990909-d2996bea-3162-4915-941d-944773505be4.png">
+
+`RPC` started with `XML` and later `JSON`-based versions.
+However, in 2015, Google created [`gRPC`](https://grpc.io/),
+a general-purpose `RPC` framework.
+Systems like Facebok's [`Apache Thrift`](https://thrift.apache.org/)
+or Twitch's [`Twirp`](https://blog.twitch.tv/en/2018/01/16/twirp-a-sweet-new-rpc-framework-for-go-5f2febbf35f/)
+use `gRPC` for internal communication between microservices.
+
+In these services, as *millions* of calls are made per-day,
```
Comment (nelsonic) [inline README.md:140]: This is a _very_ long sentence. 

image

Please avoid writing sentences with more than 25 words as it dramatically lowers readability.
The average person would totally lose track of what they are reading half way into reading this.
Not saying that the people reading this doc are "stupid", they obvs aren't!
It's a highly technical doc aimed at an educated audience.
But just so you get into the habit of improving the readability of your writing for _all_ areas of your communication. Write shorter sentences. Thank me later. 👌

**Review by nelsonic — APPROVED** [review]
@LuchoTurtle thank you very much for taking the time to do this 2023 refresh/update! 🎉 
Lots of great new info and good to get it re-read by someone with fresh eyes and enhanced! 🚀 

Still a few things we can add - will open a new issue - but this is an _excellent_ update. Thanks again! 🙏

**Pull request merged**